core/src/main/java/io/grpc/ServerImpl.java

```diff
@@ -249,6 +257,7 @@ public void transportTerminated() {
     @Override
     public ServerStreamListener streamCreated(final ServerStream stream, final String methodName,
         final Metadata.Headers headers) {
+      final Future<?> timeoutFuture = scheduleTimeout(stream, headers);
```

Comment:
I moved this to ServerStreamListenerImpl, though I am not sure about threading issues (doesn't seem like there are any?). 

There are some optimizations that could be done around not even starting the rpc if the timeout has occurred, (checking timeout.isDone()), but I am not sure its needed for correctness.   At best it would ease scenarios where the server is overloaded.